### ISSUE TITLE
Replace browser field with unpkg field in package.json.

### ISSUE THREAD

**Issue by mbostock** [d3/d3#3138]
Per unpkg/unpkg-website#63.

- [x] [d3](https://github.com/d3/d3/blob/master/package.json)

Core Modules:

- [x] [d3-array](https://github.com/d3/d3-array/blob/master/package.json)
- [x] [d3-axis](https://github.com/d3/d3-axis/blob/master/package.json)
- [x] [d3-brush](https://github.com/d3/d3-brush/blob/master/package.json)
- [x] [d3-chord](https://github.com/d3/d3-chord/blob/master/package.json)
- [x] [d3-collection](https://github.com/d3/d3-collection/blob/master/package.json)
- [x] [d3-color](https://github.com/d3/d3-color/blob/master/package.json)
- [x] [d3-dispatch](https://github.com/d3/d3-dispatch/blob/master/package.json)
- [x] [d3-drag](https://github.com/d3/d3-drag/blob/master/package.json)
- [x] [d3-dsv](https://github.com/d3/d3-dsv/blob/master/package.json)
- [x] [d3-ease](https://github.com/d3/d3-ease/blob/master/package.json)
- [x] [d3-force](https://github.com/d3/d3-force/blob/master/package.json)
- [x] [d3-format](https://github.com/d3/d3-format/blob/master/package.json)
- [x] [d3-geo](https://github.com/d3/d3-geo/blob/master/package.json)
- [x] [d3-hierarchy](https://github.com/d3/d3-hierarchy/blob/master/package.json)
- [x] [d3-interpolate](https://github.com/d3/d3-interpolate/blob/master/package.json)
- [x] [d3-path](https://github.com/d3/d3-path/blob/master/package.json)
- [x] [d3-polygon](https://github.com/d3/d3-polygon/blob/master/package.json)
- [x] [d3-quadtree](https://github.com/d3/d3-quadtree/blob/master/package.json)
- [x] [d3-queue](https://github.com/d3/d3-queue/blob/master/package.json)
- [x] [d3-random](https://github.com/d3/d3-random/blob/master/package.json)
- [x] [d3-request](https://github.com/d3/d3-request/blob/master/package.json)
- [x] [d3-scale](https://github.com/d3/d3-scale/blob/master/package.json)
- [x] [d3-selection](https://github.com/d3/d3-selection/blob/master/package.json)
- [x] [d3-shape](https://github.com/d3/d3-shape/blob/master/package.json)
- [x] [d3-time-format](https://github.com/d3/d3-time-format/blob/master/package.json)
- [x] [d3-time](https://github.com/d3/d3-time/blob/master/package.json)
- [x] [d3-timer](https://github.com/d3/d3-timer/blob/master/package.json)
- [x] [d3-transition](https://github.com/d3/d3-transition/blob/master/package.json)
- [x] [d3-voronoi](https://github.com/d3/d3-voronoi/blob/master/package.json)
- [x] [d3-zoom](https://github.com/d3/d3-zoom/blob/master/package.json)

Non-Core Modules:

- [x] [d3-contour](https://github.com/d3/d3-contour/blob/master/package.json)
- [x] [d3-fetch](https://github.com/d3/d3-fetch/blob/master/package.json)
- [x] [d3-geo-projection](https://github.com/d3/d3-geo-projection/blob/master/package.json)
- [x] [d3-hcg](https://github.com/d3/d3-hcg/blob/master/package.json)
- [x] [d3-hexbin](https://github.com/d3/d3-hexbin/blob/master/package.json)
- [x] [d3-hsv](https://github.com/d3/d3-hsv/blob/master/package.json)
- [x] [d3-require](https://github.com/d3/d3-require/blob/master/package.json)
- [x] [d3-sankey](https://github.com/d3/d3-sankey/blob/master/package.json)
- [x] [d3-scale-chromatic](https://github.com/d3/d3-scale-chromatic/blob/master/package.json)
- [x] [d3-selection-multi](https://github.com/d3/d3-selection-multi/blob/master/package.json)
- [x] [d3-tile](https://github.com/d3/d3-tile/blob/master/package.json)

Non-D3 Modules:

- [x] [topojson/topojson](https://github.com/topojson/topojson/blob/master/package.json)
- [x] [topojson/topojson-client](https://github.com/topojson/topojson-client/blob/master/package.json)
- [x] [topojson/topojson-server](https://github.com/topojson/topojson-server/blob/master/package.json)
- [x] [topojson/topojson-simplify](https://github.com/topojson/topojson-simplify/blob/master/package.json)
- [x] [mbostock/shapefile](https://github.com/mbostock/shapefile/blob/master/package.json)
- [x] [mbostock/stream-source](https://github.com/mbostock/stream-source/blob/master/package.json)
- [x] [mbostock/path-source](https://github.com/mbostock/path-source/blob/master/package.json)
- [x] [mbostock/slice-source](https://github.com/mbostock/slice-source/blob/master/package.json)
- [x] [mbostock/file-source](https://github.com/mbostock/file-source/blob/master/package.json)
- [x] [mbostock/array-source](https://github.com/mbostock/array-source/blob/master/package.json)

Other:

- [x] [Let’s Make a (D3) Plugin](https://bost.ocks.org/mike/d3-plugin/)

**Comment by jimaek** [comment]
How about using a neutral name like "cdn" or anything else really, since unpkg is not the only one that would benefit from it?
For example jsDelivr has the same npm-proxy function 
https://www.jsdelivr.com/package/npm/d3
https://github.com/jsdelivr/jsdelivr#usage

**Comment by mbostock** [comment]
I’m not responsible for the name; unpkg reads `unpkg`, `browser`, `main`. I could add the `cdn` or `jsdelivr` field simultaneously, I suppose.

**Comment by jimaek** [comment]
Well adding `jsdelivr` as well would help a lot and make all your modules work correctly on jsDelivr as well. If thats possible then its the second best option

**Comment by mbostock** [comment]
The danger of a neutral name like `cdn` is that there isn’t sufficient community consensus on what that name means—that was the problem with the `browser` field. Maybe `cdn` is specific enough to avoid that problem, but maybe it’s better to use even more specific names like `jsdelivr` and `unpkg`. But related to this, it isn’t practical for me to add explicit support for every content distribution / packaging system in the world, and I’ve been explicit since D3 4.0 that I now only support GitHub, npm, d3js.org and unpkg.com releases. So, another option would be for jsDelivr to support reading the `unpkg` field, and for unpkg to support reading the `jsdelivr` field.

**Comment by mbostock** [comment]
So, good news. It appears 45 of the 53 tasks above are unnecessary because they just use the `main` field: the UMD is suitable for both Node and the browser.

**Comment by jimaek** [comment]
Is there anything we can do to get added to the supported list of d3? The project gets over 33million CDN hits per month at the moment.
We technically could support the unpkg field but its basically brand promotion. `cdn` would be much better but I understand that its not your call

**Comment by mbostock** [comment]
Understood. I think you could do it silently without advertising support, while also working together with unpkg to establish the `cdn` as the standard.

**Comment by jimaek** [comment]
Well lets see how it goes, maybe unpkg will agree to a neutral `cdn` tag, if they decline then we will have to decide on fallbacks.

In the meantime let me know if there is anything we can do for d3.

**Comment by mbostock** [comment]
Okay, I’ve replaced all instances of `browser` with `unpkg` and `jsdelivr`.

**Comment by jimaek** [comment]
Awesome, thank you

**Comment by bryanculver** [comment]
Just as a heads up, D3 changing to meet the RFC broke https://github.com/Shopify/schmooze for me last night. As a temporary fix I locked at `4.10.0` (instead of `^4.10.0` as I was using before).

I'm still trying to narrow down the path to the issue, however `d3-request` depends on `xmlhttprequest` which in turn loads a few modules such as `child_process, fs, http` that all don't seem to have shims for browsers. I'm guessing somewhere in sprockets/schmooze path it's picking up on the `browser` section of the package.json.

I'll update/submit a separate issue once I find the responsible place to report it. For information sake, I'm not doing anything fancy, just have d3 required in `package.json` in my fairly vanilla Rails 5.1.1 installation.

**Comment by mbostock** [comment]
You shouldn’t need those shims for browsers—they are only needed for Node. Browsers have XMLHttpRequest built-in.

**Comment by bryanculver** [comment]
Yeah, so I thought myself. I have no idea why it cropped up but as soon as I go back to 4.10.0, all is well. I have clean/upgraded/reinstalled to no avail on 4.10.1. One thing I am doing is running through babel because I have a deployment targets with older browsers.

I'm not suggesting this need to be rolled back, I like the direction it is going. However I know someone else is bound to stumble upon this as well.

**Comment by mbostock** [comment]
My guess is that schmooze gave the `browser` entry higher priority than `main`, and now that that’s gone from d3-request 1.0.6 (d3 4.10.1), it is pulling in the code that is intended for Node rather than the browser. Unfortunately, I don’t support this use case. The supported environments are Node (via npm and CommonJS require), browsers (via npm-backed CDNs and UMD, and via the zip attached to each release posted on GitHub) and ES modules (via npm and the `module` entry). If you are using a bundler targeting the web, you should use a bundler that consumes the ES modules such as Rollup or webpack.

**Comment by bryanculver** [comment]
No worries! My comment was geared toward fellow head-scratchers and wanting to point them into the direction of "What changed?"

As for my project, "Move to Webpack" is on my list of things to do.

Happy coding!

**Comment by mbostock** [comment]
I don’t actually see any module resolution in schmooze, so I am guessing maybe you wrote own code to find the JavaScript file to pass to schmooze? So you could perhaps update that code to consume the browser-targeting UMD bundle directly rather than using the `main` entry in the package.json.

**Comment by akbansal** [comment]
So I am am getting the same issue.. when D3 changed the version from x.10.0 to x.10.2.. now in our project we are using webpack and babel. so how can I define d3 in package.json or webpack.config to use d3 latest version? any example?

**Comment by mbostock** [comment]
@akbansal Please see d3/d3-request#24.

**Comment by akbansal** [comment]
Thanks @mbostock, now issue is resolved with latest D3 version (4.10.2). below is the change I did in my webpack.config file.. (just in case if someone else also looking for the solution - 

```
Package.json (no change in package.json)
   "devDependencies": {
          ...
         "d3": "^4.4.2",
         ...
   }

in webpack.config file - 

    resolve: {
          extensions: ['', '.js', '.vue'],
          fallback: [path.join(__dirname, '../node_modules')],
          alias: {
               'src': path.resolve(__dirname, '../src'),
               'assets': path.resolve(__dirname, '../src/assets'),
               'components': path.resolve(__dirname, '../src/components'),
               **'d3': 'd3/build/d3.js'**
    }
```

Thanks @mbostock

**Comment by macalinao** [comment]
Fixed worked for me @akbansal. Thanks!
